### PR TITLE
Python deriv api/example cache

### PR DESCRIPTION
**Pull request by sasikala-binary**

### CODE REVIEW

File: deriv_api/deriv_api.py [hunk 1]
```diff
@@ -1,48 +1,123 @@
...
+    param {String}     options.lang       - Language of the API communication
+    param {String}     options.brand      - Brand name
+    param {Object}     options.middleware - A middleware to call on certain API actions
+
+    property {Cache} cache - Temporary cache default to {InMemory}
+    property {Cache} storage - If specified, uses a more persistent cache (local storage, etc.)
+    """
     wsconnection:str = ''
+    storage = ''
+    def __init__(self, options):
```
Comment (chylli-binary) [inline deriv_api/deriv_api.py:38]: Maybe 
```suggestion
    def __init__(self, **options):
```
will be better ?

File: deriv_api/utils.py [hunk 2]
```diff
@@ -1,9 +1,19 @@
...
+
+def is_valid_url(url):
+    regex = re.compile(
+        r'^(?:ws)s?://'  # ws:// or wss://
+        r'(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|'  # domain...
+        r'localhost|'  # localhost...
+        r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})'  # ...or ip
+        r'(?::\d+)?'  # optional port
+        r'(?:/?|[/?]\S+)$', re.IGNORECASE)
+    return re.match(regex, url) is not None
```
Comment (chylli-binary) [inline deriv_api/utils.py:19]:
```suggestion
    return re.match(regex, url) is not None

```

File: tests/test_deriv_api.py [hunk 3]
```diff
@@ -14,3 +30,16 @@ def test_deriv_api(mocker):
...
+    assert deriv_api_obj.get_url("localhost") == "wss://localhost"
+    assert deriv_api_obj.get_url("ws://localhost") == "ws://localhost"
+    with pytest.raises(ConstructionError, match=r"Invalid URL:testurl"):
+        deriv_api_obj.get_url("testurl")
+
+
+def get_deriv_api(mocker):
+    mocker.patch('deriv_api.deriv_api.DerivAPI.api_connect', return_value='')
+    deriv_api_obj = deriv_api.DerivAPI({'app_id': 1234, 'endpoint': 'localhost'})
+    return deriv_api_obj
```
Comment (chylli-binary) [inline tests/test_deriv_api.py:45]:
```suggestion
    return deriv_api_obj

```

**Comment by chylli-binary** [comment]
Looks good for me.

File: deriv_api/deriv_api.py [hunk 4]
```diff
@@ -1,48 +1,123 @@
...
 from deriv_api.deriv_api_calls import DerivAPICalls
+from deriv_api.in_memory import InMemory
 import websockets
 import json
 import logging
+from deriv_api.errors import APIError, ConstructionError
+from deriv_api.utils import dict_to_cache_key, is_valid_url
+import re
 
 # TODO: remove after development
```
Comment (shantanu-deriv) [inline deriv_api/deriv_api.py:11]:
```suggestion
```

**Review by shantanu-deriv — APPROVED** [review]

**Pull request merged**